packages/fxa-metrics-processor/src/lib/amplitude/transforms/amplitude-identify.ts

```diff
@@ -0,0 +1,100 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { AmplitudeHttpEvent, EventContext } from './types';
+import { ServiceNameAndClientIdMapper, toSnakeCase } from './common';
+
+// See https://developers.amplitude.com/?javascript#identify-api
+
+// Event though we include all these verbs here, only $append is used in practice.
+const IDENTIFY_VERBS = ['$set', '$setOnce', '$add', '$append', '$prepend', '$unset'] as const;
+type IdentifyVerb = typeof IDENTIFY_VERBS[number];
+export type AmplitudeIdentifyEvent = {
+  device_id?: AmplitudeHttpEvent['device_id'];
+  event_type: '$identify';
+  user_id: AmplitudeHttpEvent['user_id'];
+  user_properties: IdentifyEventUserProperties;
+};
+type IdentifyEventUserProperties = {
+  [key in Exclude<IdentifyVerb, '$append'>]?: { [key: string]: {} };
+} & {
+  $append?: AppendUserProperties;
+};
+type AppendUserProperties = {
+  [key: string]: string | string[];
+};
+
+/**
+ * Create the payload for an Amplitude "identify" event.  Currently this _only_
+ * uses `$append` because historically that's what FxA services did.
+ *
+ * @param context Raw event context
+ * @param amplitudeEventHttpEvent The Amplitude http/batch API payload
+ *                                associated with the identify event
+ * @param servicePropMapper A function for mapping service names and client ids
+ */
+export function createAmplitudeIdentityEvent(
+  context: EventContext,
+  amplitudeEventHttpEvent: AmplitudeHttpEvent,
+  servicePropMapper: ServiceNameAndClientIdMapper
+): AmplitudeIdentifyEvent | null {
+  const servicesUsed = mapServicesUsed(servicePropMapper, context);
+  const experiments = mapExperiments(context);
+  const userPreferences = mapUserPreferences(context);
+
+  if (servicesUsed || experiments || userPreferences) {
+    return {
+      device_id: amplitudeEventHttpEvent.device_id,
+      event_type: '$identify',
+      user_id: amplitudeEventHttpEvent.user_id,
+      user_properties: {
+        $append: {
+          ...servicesUsed,
+          ...experiments,
+          ...userPreferences,
+        },
+      },
+    };
+  }
+
+  return null;
+}
+
+function mapServicesUsed(servicePropMapper: ServiceNameAndClientIdMapper, context: EventContext) {
```

Comment:
General typescript question, function above specifies null and AmplitudeIdentifyEvent as a return type and others don't specify anything even though they do return null. Is there a standard or best practice for this?

Reply:
Off hand I don't have a link or something to share for this.  fxa-metrics-processor has a pretty strict `tsconfig.json`, which among other things, does not allow implicit `any`s.  So as far as tsc doesn't complain about the inferred types, I think it's ok.

The other reasons I didn't type them are that they are not exported, they are very simple types, and iirc I wouldn't have used them anywhere else aside as the return types.